File: javascript/easy/movingAverageFromDataStream.js
/**
 * Initialize your data structure here.
 * @param {number} size
 */
var MovingAverage = function(size) {
    this.queue = [];
    this.sum = 0;
    this.size = size;
};

/**
 * @param {number} val
 * @return {number}
 */
MovingAverage.prototype.next = function(val) {
    let queue = this.queue, size = this.size;
    if(queue.length >= size) {
        this.sum -= queue.shift();
    }
    queue.push(val);
    this.sum += val;
    return (this.sum/queue.length);
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * var obj = Object.create(MovingAverage).createNew(size)
 * var param_1 = obj.next(val)
 */

File: ruby/easy/moving_average_from_data_stream.rb
class MovingAverage

=begin
    Initialize your data structure here.
    :type size: Integer
=end
    def initialize(size)
        @size = size
        @queue = Array.new()
        @sum = 0
    end


=begin
    :type val: Integer
    :rtype: Float
=end
    def next(val)
        if(@queue.length >= @size)
            @sum -= @queue.shift
        end
        @queue.push(val)
        @sum += val
        return (@sum.to_f/@queue.length)
    end


end

# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage.new(size)
# param_1 = obj.next(val)
